### DISCUSSION TITLE
2. August 2022

### DISCUSSION THREAD

**Discussion by dimakuv** [gramineproject/gramine#792] [Meeting notes]
(**Please note that the meetings happen on Tuesdays from now on!**)

### Agenda

*(please write your proposed agenda items in comments under this discussion)*

- [x] Quick announcement: Dmitrii (@dimakuv) goes on vacation from 8. August to 15. August, but can check emails periodically.
- [x] Status update on `loader.argv = ["arg0", "arg1", "arg2"]` (https://github.com/gramineproject/gramine/issues/761).
- [x] Status update on sanitization of `/etc/` files (https://github.com/gramineproject/gramine/issues/689)
- [x] Status update on Microsoft Azure Attestation (https://github.com/gramineproject/gramine/issues/626)
- [x] `LD_LIBRARY_PATH` handling in Gramine, see discussion at https://github.com/gramineproject/gramine/pull/781#pullrequestreview-1050717986

### Meeting Notes for 9. August

Dmitrii is on vacation, so Vijay (@vijaydhanraj) will keep the meeting notes for the 9. August meeting.

### Status update on `loader.argv = ["arg0", "arg1", "arg2"]`

No work is done currently. Mariusz (@oshogbo) was preempted to work on the "sanitization of `/etc/ files`", and didn't have time to look into this issue.

### Status update on sanitization of `/etc/` files

Mariusz is actively working on this. See the proposal details here: https://github.com/gramineproject/gramine/issues/689#issuecomment-1201048997

### Status update on Microsoft Azure Attestation

Dmitrii will present the MAA support in the meeting on 16. August. Ideally, we'd like to start the review process soon. The details about MAA were summarized here: https://github.com/gramineproject/gramine/issues/626

The MAA PR is here: https://github.com/gramineproject/gramine/pull/652. Note that there are **no** changes to the core Gramine, only to the RA-TLS library.

- *Woju* asked how to test this in our Jenkins CI? Ideally, Microsoft Azure could allocate some VMs for us, for CI purposes.
    - **TODO(@monavij)**: send a request to Microsoft about getting an Azure CC VM.

### `LD_LIBRARY_PATH` handling in Gramine

The context is in the comments for this PR: https://github.com/gramineproject/gramine/pull/781

Gramine relies on a hacky trick in the manifest file: users are required to override `LD_LIBRARY_PATH` such that the first path points to the Gramine-specific patched Glibc. See for example: https://github.com/gramineproject/gramine/blob/f91a2c18370858701a33c00620b0c7954f6779bf/CI-Examples/redis/redis-server.manifest.template#L32

This is hacky but usually works, since normal environments do not set `LD_LIBRARY_PATH` and do not play tricks with this envvar. However, Woju experienced problems with it: if the native app runs with a custom `LD_LIBRARY_PATH`, then Gramine manifest needs to reflect the custom `LD_LIBRARY_PATH` paths when overriding it. Also, Borys expects some Bash scripts to blow up, because sometimes Bash scripts change `LD_LIBRARY_PATH` and then execute some app -- Gramine's child process will override that custom `LD_LIBRARY_PATH` with the default override, and standard libs may go out of sync.

Proposals:
1. Woju: add the manifest syntax with "explicit file contents". Then the user may set up her own `/etc/ld.so.conf` contents that include our patched-Glibc `/lib/:` entry.
2. Borys: *nuclear solution* -- add a manifest option `use_our_patched_glibc = true`, and this option will reconfigure ~`LD_LIBRARY_PATH` and~ the corresponding in-Gramine FS files so that host Glibc is completely shadowed by our patched Glibc. The downside of this solution is that it does a lot of unrelated things, feels like black magic, and doesn't provide any flexbility for the manifest writer (and may lead to weird side effects in the manifest contains contradicting entries).
3. Borys: shadow (remove) `/etc/ld.so.cache` (output of `/etc/ld.so.conf` compilation) in Gramine-FS, then we can rely on the `/lib` (or `/lib64`, need to verify) having priority over other paths (e.g. `/usr/lib`). We would use that directory in all of our examples and remove `LD_LIBRARY_PATH` from the manifests. The only caveat is that we either need to ask users to shadow/remove/not mount `/etc/ld.so.cache` in their manifests (to be fair they shouldn't do that even now) or do it implicitly (I don't like this option - surprising to the user).

**Comment by vans163** [comment]
Entry into iDash, task 3 specifically, looks like a good challenge for the /dev/shm addition coming soon to gramine. http://www.humangenomeprivacy.org/2022/competition-tasks.html

**Reply by boryspoplawski** [reply]
Am I the only one that finds it ironic that a site with "privacy" in its domain doesn't support https?

**Reply by vans163** [reply]
The site seems to have many contributors/organisers chinese and from inside mainland china firewall https barely works, but that is no excuse as most devs have custom proxies/vpns anyways.

**Comment by vans163** [comment]
Question:  

How safe is /dev/sgx to be given to a non-sudo untrusted linux user. Our usecase is we want to (for KISS) expose /dev/sgx into a sandboxed non-root podman container that has memory and CPU limits set via cgroupsv2.  If the arbitrary untrusted user of said container starts a SGX enclave can they exceed the memory limits of cgroupsv2?  Cpu limits? Also is it safe to do this. 

Option 2 would be to write a wrapper API that will execute /dev/sgx with the requested parameters, but this is less flexibility for the end user and (much) more coding work for us.

**Reply by kailun-qin** [reply]
> If the arbitrary untrusted user of said container starts a SGX enclave can they exceed the memory limits of cgroupsv2? Cpu limits?

No, they can't exceed the memory (assuming you're NOT talking about the EPC memory) or CPU limits as long as they don't have arbitrary acceess/privilege to manage cgroups.

**Reply by dimakuv** [reply]
As @kailun-qin said, no, SGX enclaves cannot exceed the memory or CPU limits of cgroups.

This is because Intel SGX enclaves rely on the host OS kernel (Linux) to manage the memory and CPU resources. In particular, SGX enclave memory is actually backed by normal system memory, and SGX enclave threads are backed by normal system threads. So any system limitations like cgroups apply equally well to SGX enclaves (because from the perspective of the host OS, SGX enclaves are just normal processes).

**Reply by vans163** [reply]
@kailun-qin @dimakuv thank you!

**Reply by mkow** [reply]
From the security side option 2 would be a better idea probably, assuming you write the proxy carefully and ideally in a memory-safe language (because now you've shifted the attack surface from the SGX driver to your own code). SGX driver is relatively new and still evolving, so there is a risk that it will get vulnerable at some point (or already is). But as you said, this requires some additional work, so you need to weight the risks.

**Comment by llly** [comment]
Will this meeting be on 1st August or 2rd August?

**Reply by dimakuv** [reply]
2. August. Sorry, I forgot to update the title. I edited it now.

**Comment by kailun-qin** [comment]
> ### Status update on Microsoft Azure Attestation
> Dmitrii will present the MAA support in the meeting on 16. August. Ideally, we'd like to start the review process soon. The details about MAA were summarized here: #689

Should the details about MAA be https://github.com/gramineproject/gramine/issues/626 (instead of #689)?

**Reply by dimakuv** [reply]
My blooper. Fixed. Thanks.
